Mathlib/Algebra/Module/CharacterModule.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Junyan Xu
+-/
+
+import Mathlib.Algebra.Module.LinearMap
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.Algebra.Category.GroupCat.Injective
+import Mathlib.Topology.Instances.AddCircle
+import Mathlib.Topology.Instances.Rat
+import Mathlib.Algebra.Category.ModuleCat.ChangeOfRings
+
+/-!
+# Character module of a module
+
+For commutative ring `R` and an `R`-module `M` and an injective module `D`, its character module
+`M⋆` is defined to be `R`-linear maps `M ⟶ D`.
+
+`M⋆` also has an `R`-module structure given by `(r • f) m = f (r • m)`.
+
+## Main results
+
+- `CharacterModuleFunctor` : the contravariant functor of `R`-modules where `M ↦ M⋆` and
+an `R`-lineara map `l : M ⟶ N` induces an `R`-linear map `l⋆ : f ↦ f ∘ l` where `f : N⋆`.
+- `LinearMap.charaterfy_surjective_of_injective` : If `l` is injective then `l⋆` is surjective,
+  in another word taking character module as a functor sends monos to epis.
+- `CharacterModule.exists_character_apply_ne_zero_of_ne_zero` : for nonzero `a ∈ M`, there is a
+  character `c` in `M⋆` such that `c a` is nonzero as well.
+- `CharacterModule.homEquiv` : there is a bijection between linear map `Hom(N, M⋆)` and
+  `(N ⊗ M)⋆` given by `curry` and `uncurry`.
+
+-/
+
+open CategoryTheory
+
+universe uR uM uN uD
+variable (R : Type uR) [CommRing R]
+variable (M : Type (max uR uM)) [AddCommGroup M] [Module R M]
+variable (N : Type (max uR uN)) [AddCommGroup N] [Module R N]
+variable (D : Type (max uR uD)) [AddCommGroup D] [Module R D]
+variable [injective_dual : Injective <| ModuleCat.of R D]
+
+-- we really want to consider only injective modules
+/--
+If `M` is an `R`-module, its `D`-character module is defined to be the `Hom_R(M, D)`
+-/
+@[nolint unusedArguments]
+def CharacterModule [Injective <| ModuleCat.of R D] : Type (max uR uM uD) := M →ₗ[R] D
+
+instance : FunLike (CharacterModule.{uR, uM, uD} R M D) M (fun _ ↦ D) :=
+  inferInstanceAs <| FunLike (M →ₗ[R] D) M _
+
+instance : AddCommGroup (CharacterModule.{uR, uM, uD} R M D) :=
+  inferInstanceAs <| AddCommGroup <| M →ₗ[R] D
+
+instance : Module R (CharacterModule.{uR, uM, uD} R M D) :=
+  LinearMap.module
+
+instance : LinearMapClass (CharacterModule.{uR, uM, uD} R M D) R M D :=
+  inferInstanceAs <| LinearMapClass (M →ₗ[R] D) R M D
+
+@[simp] lemma CharacterModule.smul_apply (f : CharacterModule.{uR, uM, uD} R M D) (r : R) (m : M) :
+    (r • f) m = f (r • m) := by
+  rw [LinearMap.smul_apply, f.map_smul]
+
+variable {R M N}
+
+/--
+For a linear map `L : M → N`, `(· ∘ L)` defines map from `CharacterModule N` to `CharacterModule M`
+-/
+@[simps]
+def LinearMap.characterify
+    (L : M →ₗ[R] N)  :
+    CharacterModule.{uR, uN, uD} R N D →ₗ[R] CharacterModule.{uR, uM, uD} R M D where
```

Comment:
This one can be defined using `flip` and `llcomp` just like [Module.Dual.transpose](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Dual.html#Module.Dual.transpose), at least in the commutative case. The name could be characterDual in analogy to [LinearMap.dualMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Dual.html#LinearMap.dualMap).